### ISSUE TITLE
iOS contact database not found - potential fix in description

### ISSUE THREAD

**Issue by achalddave** [KnugiHK/WhatsApp-Chat-Exporter#76]
Thanks for this great tool! I received the following error when using this with the latest version of WhatsApp: "Contact database not found." I think the uuid(?) for the contacts db in whatsapp may have changed. 

Changing this line: https://github.com/KnugiHK/WhatsApp-Chat-Exporter/blob/c27f5ee41cd97e5b954cff4dc4c08ba870be0fbf/Whatsapp_Chat_Exporter/utility.py#L299

from
```
    CONTACT = "b8548dc30aa1030df0ce18ef08b882cf7ab5212f"
```

to
```
    CONTACT = "016df2e52e5bfaa000f98a1dbe5c40cd6dc7f657"
```

fixed the issue for me. I'm not sure if this is a general fix, so I'm not submitting a PR, but opening this issue in case anyone else runs into it.

**Comment by KnugiHK** [comment]
I currently don't have access to the latest iOS WhatsApp. Once it has been confirmed, I will make changes to the code. Thanks for your report and the potential fix anyway!

**Comment by SvenHerr** [comment]
@achalddave how did you know that this is the new db ? i cant find 016df2e52e5bfaa000f98a1dbe5c40cd6dc7f657 with the version 17.2.1

**Comment by achalddave** [comment]
I'm sure there's a better way, but I went to the path to my Manifest.db, which was in a path like: `~/Library/Application Support/MobileSync/Backup/<numbers>/Manifest.db`, and then ran:

```
~ sqlite3 Manifest.db
SQLite version 3.39.5 2022-10-14 20:58:05
Enter ".help" for usage hints.
sqlite> select * from files;
```

This printed a bunch of tables, and I looked for one which said `ContactsV2.sqlite`, and copied the hash at the start of that line.

**Comment by KnugiHK** [comment]
> I'm sure there's a better way, but I went to the path to my Manifest.db, which was in a path like: `~/Library/Application Support/MobileSync/Backup/<numbers>/Manifest.db`, and then ran:
> 
> ```
> ~ sqlite3 Manifest.db
> SQLite version 3.39.5 2022-10-14 20:58:05
> Enter ".help" for usage hints.
> sqlite> select * from files;
> ```
> 
> This printed a bunch of tables, and I looked for one which said `ContactsV2.sqlite`, and copied the hash at the start of that line.

This is the correct way to find a file inside the iOS backup.

**Comment by SvenHerr** [comment]
thank you for your reply. I just looked throu all files with .sqlite and i cant find any db with contacts.

**Comment by Rerebla** [comment]
I get the same error "Contact database not found". Running the program with the default or the suggested fix results in the same error. I couldn't find any ContactsV2 or anything containing 'contacts' related to WhatsApp.

**Comment by roisec** [comment]
@angelbanderasudg it's solved the issue but after I have another issue
Unexpected contact format: broadcast
```
  File "/Users//repos/iphone/lib/python3.11/site-packages/Whatsapp_Chat_Exporter/__main__.py", line 425, in main
    create_html(
  File "/Users//repos/iphone/lib/python3.11/site-packages/Whatsapp_Chat_Exporter/android_handler.py", line 753, in create_html
    safe_file_name, name = get_file_name(contact, chat)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users//repos/iphone/lib/python3.11/site-packages/Whatsapp_Chat_Exporter/utility.py", line 141, in get_file_name
    raise ValueError("Unexpected contact format: " + contact)
ValueError: Unexpected contact format: broadcast
```

**Comment by jpfleischer** [comment]
please merge @angelbanderasudg 's fix

**Comment by KnugiHK** [comment]
> @angelbanderasudg it's solved the issue but after I have another issue Unexpected contact format: broadcast
> 
> ```
>   File "/Users//repos/iphone/lib/python3.11/site-packages/Whatsapp_Chat_Exporter/__main__.py", line 425, in main
>     create_html(
>   File "/Users//repos/iphone/lib/python3.11/site-packages/Whatsapp_Chat_Exporter/android_handler.py", line 753, in create_html
>     safe_file_name, name = get_file_name(contact, chat)
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users//repos/iphone/lib/python3.11/site-packages/Whatsapp_Chat_Exporter/utility.py", line 141, in get_file_name
>     raise ValueError("Unexpected contact format: " + contact)
> ValueError: Unexpected contact format: broadcast
> ```

I notice that your handler is for Android. Please open a separate issue for that as this issue exclusively appear in iOS.

**Comment by KnugiHK** [comment]
A workaround has been implemented in #106.

**Comment by KnugiHK** [comment]
Potential fix proposed in https://github.com/KnugiHK/WhatsApp-Chat-Exporter/issues/109#issuecomment-2267853193 by @ReclusiveEagle.

**Comment by KnugiHK** [comment]
> I currently don't have access to the latest iOS WhatsApp. Once it has been confirmed, I will make changes to the code. Thanks for your report and the potential fix anyway!

I created an iOS (17.6.1) backup with WhatsApp 2.24.17.78. I do see the ContactsV2.sqlite file. Not sure how to reproduce the issue. Any ideas?

**Comment by KnugiHK** [comment]
Hi @ReclusiveEagle, I took a look at your solution and it does not seem right to me.

> Using any of the unique hashes assigned to anyone of your contact's chat seem to fix the issue.
>
> For example:
>
>0618655a1de1254eca90c229002cdea325de408d for:
>```
>0618655a1de1254eca90c229002cdea325de408dAppDomainGroup-group.net.whatsapp.WhatsApp.sharedMessage/Media/[Contacts Number]@s.whatsapp.net/0/6/06693485-42bb-4f27-844d-1c37d93c1291.opusbplist00
>```
> In utility.py, for class WhatsAppIdentifier(StrEnum): replace the hash in CONTACT = "to the unique hash from one of your contacts." and run the script again.

However, the contact file should be a SQLite database, while your solution seems to be extracting a non-SQLite file—specifically, an audio file in this case.

**Comment by jpfleischer** [comment]
for whatever reason, the hash from the audio file works. this has allowed me to use this library without throwing any errors

**Comment by KnugiHK** [comment]
> for whatever reason, the hash from the audio file works. this has allowed me to use this library without throwing any errors

That might indeed help resolving the contact database not found error (I haven't test it), but it's not really a proper solution.

Instead, the contact database could treated as optional, if iOS backups can't guarantee its inclusion.

**Comment by KnugiHK** [comment]
The commit fe88f1b should allow the tool to proceed even if the database is not presented.

**Comment by KnugiHK** [comment]
The contact database may not be included in the iOS backup. Since there are no alternative files to replace it, the commit fe88f1b is deemed an appropriate solution. This fix was released in 0.10.5.
